### PR TITLE
Admission Webhooks for OLM

### PR DESCRIPTION
**Pull request by ecordell**
**Description of the change:**
This adds an admission webhook to OLM that:

- Detects when a user is creating a CSV
- Processes the CSV for the required RBAC
- Checks if the user has enough permission to have created the RBAC directly
- If so, creates the RBAC for the CSV

Directly applying a CSV and CRD will result in the CSV transitioning past `Pending` and into running states.

If the RBAC generation fails for some reason, the CSV status will explain exactly which rbac rules still need to be filled.

The processing is done asynchronously and doesn't block storage of the CSV by etcd.

The webhook requires TLS certs coordinated with the kube apiserver. For this PR, the webhook is only enabled when running locally, and certs are generated at that time. Upstream deployments should probably integrate with cert-manager, and OpenShift deployments should use service-ca-signer (this work is coming).

**Motivation for the change:**

It can be frustrating to iterate on CSVs if you have to syncronize rbac between CSV and standard roles/rolebindings.

This also restores the older behavior of OLM, which 

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/controller/operators/olm/admission/config.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+// WithPermissionValidator sets the permission validator used by the RBAC generator.
+func WithPermissionValidator(validator permissions.Validator) RbacGenerationOption {
+	return func(config *rbacGenerationConfig) {
+		config.userPermissionValidator = validator
+	}
+}
+
+// WithPermissionValidator sets the permission validator used by the RBAC generator.
```
Comment (njhale) [inline pkg/controller/operators/olm/admission/config.go]:
```suggestion
// WithPermissionCreator sets the permission validator used by the RBAC generator.
```

File: deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml [hunk 2]
```diff
@@ -85,17 +88,22 @@ spec:
           resources:
 {{ toYaml .Values.olm.resources | indent 12 }}
           {{- end}}
-          {{ if and .Values.installType (eq .Values.installType "ocp") }}
           volumeMounts:
+          {{ if and .Values.installType (eq .Values.installType "ocp") }}
           - mountPath: /var/run/secrets/serving-cert
```
[the remark this answers is not in the document]
Comment (ecordell) [reply deploy/chart/templates/0000_50_olm_07-olm-operator.deployment.yaml]: The service-ca-signer doesn't support injecting cabundles into validating admission webhooks - I don't think we have a good solution for openshift right now.

File: deploy/chart/templates/0000_50_olm_20-admission.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+{{- $ca := genCA "admission-ca" 365 -}}
```
[the remark this answers is not in the document]
Comment (ecordell) [reply deploy/chart/templates/0000_50_olm_20-admission.yaml]: It's the local solution, and I'm thinking we should repurpose this PR as a "local-only" feature for now, and follow up with the upstream/downstream versions.

File: deploy/chart/templates/0000_50_olm_20-admission.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  name: olm-admission-certs
+  namespace: {{ .Values.namespace }}
+data:
+  {{- $altNames := list ( printf "olm-admission.%s" .Values.namespace ) ( printf "olm-admission.%s.svc" .Values.namespace ) -}}
+  {{- $cert := genSignedCert "olm-admission" nil $altNames 365 $ca }}
+  tls.crt: {{ $cert.Cert | b64enc | quote }}
+  tls.key: {{ $cert.Key | b64enc | quote }}
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
```
[the remark this answers is not in the document]
Comment (ecordell) [reply deploy/chart/templates/0000_50_olm_20-admission.yaml]: MutatingWebhooks have a different lifecycle - this webhook isn't mutating the object itself so I think validating makes sense for now (in the future we may decide we want the other lifecycle?)

File: deploy/chart/templates/0000_90_olm_00-service-monitor.yaml [hunk 5]
```diff
@@ -36,7 +36,6 @@ metadata:
   labels:
     app: catalog-operator
 spec:
-  jobLabel: k8s-app
```
[the remark this answers is not in the document]
Comment (ecordell) [reply deploy/chart/templates/0000_90_olm_00-service-monitor.yaml]: yes - it's a duplicated key

File: deploy/chart/templates/0000_50_olm_20-admission.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  name: olm-admission
+webhooks:
+- name: csv.operators.coreos.com
+  rules:
+    - apiGroups:
+        - "operators.coreos.com"
+      apiVersions:
+        - v1alpha1
+      operations:
+        - CREATE
```
[the remark this answers is not in the document]
Comment (ecordell) [reply deploy/chart/templates/0000_50_olm_20-admission.yaml]: I was thinking we'd start conservatively.

File: pkg/admission/server.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+	klog.V(2).Info(fmt.Sprintf("handling request: %s", body))
+
+	// The AdmissionReview that was sent to the webhook
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+
+	// The AdmissionReview that will be returned
+	responseAdmissionReview := v1beta1.AdmissionReview{}
+
+	deserializer := codecs.UniversalDeserializer()
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/admission/server.go]: I intend to follow up with updates to this - there was some issue integrating with the current stack that I can't recall right now.

File: og.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+metadata:
+  name: olmadmin-minikube
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: olmadmin
+subjects:
+- kind: User
+  name: minikube-user
+
```
[the remark this answers is not in the document]
Comment (ecordell) [reply og.yaml]: yep! removed.

File: pkg/admission/server.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+func (c *Config) addFlags() {
+	flag.StringVar(&c.CertFile, "tls-cert-file", c.CertFile, ""+
+		"File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated "+
+		"after server cert).")
+	flag.StringVar(&c.KeyFile, "tls-private-key-file", c.KeyFile, ""+
+		"File containing the default x509 private key matching --tls-cert-file.")
+}
+
+func configTLS(config Config) *tls.Config {
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/admission/server.go]: this is aligning with the upstream webhook example

**Comment by awgreene** [comment]
Closing this in favor of https://github.com/operator-framework/operator-lifecycle-manager/pull/1102

**Pull request closed without merging**